### PR TITLE
OCPBUGS-31067: server: still serve kubeconfig with node-bootstrapper-token

### PR DESCRIPTION
**Pull request by yuqi-zhang**
Reverts a recent change where we stuffed in the full kubeconfig from the apiserver-server-ca instead of using the mounted node-bootstrapper token. The caData should still be up to date, but we go back to using node-bootstrapper for node scaling so it can get the CSR auto-appoved.

### CODE REVIEW

**Comment by JoelSpeed** [comment]
Should this have changed the content of the `kubeconfig-data` configmap? As that is still using the client certificate looking at the CI runs

**Comment by yuqi-zhang** [comment]
> Should this have changed the content of the kubeconfig-data configmap? As that is still using the client certificate looking at the CI runs

That's currently used by the daemon to fetch the CA but we should probably eventually make that a secret and potentially remove the kubeconfig from there and just use it for the updated CA.

This PR just aims to add the mounted token back to the served kubeconfig file so a node can bootstrap, but still use the latest CAs from the cluster instead of the files in the MCS

**Comment by yuqi-zhang** [comment]
To make this easier to test, I cherry-picked in @cdoern 's https://github.com/openshift/machine-config-operator/pull/4274.

I believe the verification would be that the rotation still works, and on top of that, if the cluster was installed on 4.13 or earlier, they can upgrade to this PR, and new nodes are able to join the cluster without having a CSR that needs approval (conversely, if they upgraded from 4.13->4.16.ec4 without this PR, it should not automatically succeed)

**Comment by rioliu-rh** [comment]
upgrade cluster from 4.13 to the[ CI image](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/release-openshift-origin-installer-launch-aws-modern/1770976038021500928) 

upgrade is success
```console
$ cv version -o yaml | yq '.status.history[]'
{
  "acceptedRisks": "Target release version=\"\" image=\"registry.build03.ci.openshift.org/ci-ln-tyzdpr2/release:latest\" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified\nPrecondition \"ClusterVersionRecommendedUpdate\" failed because of \"UnknownUpdate\": RetrievedUpdates=False (VersionNotFound), so the recommended status of updating from 4.15.3 to 4.16.0-0.test-2024-03-22-010219-ci-ln-tyzdpr2-latest is unknown.",
  "completionTime": "2024-03-22T07:09:31Z",
  "image": "registry.build03.ci.openshift.org/ci-ln-tyzdpr2/release:latest",
  "startedTime": "2024-03-22T06:10:49Z",
  "state": "Completed",
  "verified": false,
  "version": "4.16.0-0.test-2024-03-22-010219-ci-ln-tyzdpr2-latest"
}
{
  "acceptedRisks": "Target release version=\"\" image=\"quay.io/openshift-release-dev/ocp-release:4.15.3-x86_64\" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified\nForced through blocking failures: Multiple precondition checks failed:\n* Precondition \"EtcdRecentBackup\" failed because of \"ControllerStarted\": RecentBackup: The etcd backup controller is starting, and will decide if recent backups are available or if a backup is required\n* Precondition \"ClusterVersionRecommendedUpdate\" failed because of \"UnknownUpdate\": RetrievedUpdates=False (VersionNotFound), so the recommended status of updating from 4.14.18 to 4.15.3 is unknown.",
  "completionTime": "2024-03-22T03:59:43Z",
  "image": "quay.io/openshift-release-dev/ocp-release:4.15.3-x86_64",
  "startedTime": "2024-03-22T03:05:16Z",
  "state": "Completed",
  "verified": false,
  "version": "4.15.3"
}
{
  "acceptedRisks": "Target release version=\"\" image=\"quay.io/openshift-release-dev/ocp-release:4.14.18-x86_64\" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified\nForced through blocking failures: Multiple precondition checks failed:\n* Precondition \"ClusterVersionUpgradeable\" failed because of \"AdminAckRequired\": Kubernetes 1.27 and therefore OpenShift 4.14 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6958395 for details and instructions.\n* Precondition \"EtcdRecentBackup\" failed because of \"ControllerStarted\": RecentBackup: The etcd backup controller is starting, and will decide if recent backups are available or if a backup is required\n* Precondition \"ClusterVersionRecommendedUpdate\" failed because of \"UnknownUpdate\": RetrievedUpdates=False (VersionNotFound), so the recommended status of updating from 4.13.38 to 4.14.18 is unknown.",
  "completionTime": "2024-03-22T02:51:30Z",
  "image": "quay.io/openshift-release-dev/ocp-release:4.14.18-x86_64",
  "startedTime": "2024-03-22T01:50:29Z",
  "state": "Completed",
  "verified": false,
  "version": "4.14.18"
}
{
  "completionTime": "2024-03-22T01:40:30Z",
  "image": "quay.io/openshift-release-dev/ocp-release@sha256:db7a6fad1d56cf391159d252daabbb44178132081e077ae290317b91633fb1db",
  "startedTime": "2024-03-22T01:19:19Z",
  "state": "Completed",
  "verified": false,
  "version": "4.13.38"
}
```

**Comment by rioliu-rh** [comment]
@yuqi-zhang

upgrade can be success from 4.13 to 4.16.ec.4
```console
$ cv version -o yaml | yq '.status.history[]'
{
  "acceptedRisks": "Target release version=\"\" image=\"quay.io/openshift-release-dev/ocp-release:4.16.0-ec.4-x86_64\" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified\nPrecondition \"ClusterVersionRecommendedUpdate\" failed because of \"UnknownUpdate\": RetrievedUpdates=False (VersionNotFound), so the recommended status of updating from 4.15.5 to 4.16.0-ec.4 is unknown.",
  "completionTime": "2024-03-25T04:48:52Z",
  "image": "quay.io/openshift-release-dev/ocp-release:4.16.0-ec.4-x86_64",
  "startedTime": "2024-03-25T03:52:36Z",
  "state": "Completed",
  "verified": false,
  "version": "4.16.0-ec.4"
}
{
  "acceptedRisks": "Target release version=\"\" image=\"quay.io/openshift-release-dev/ocp-release:4.15.5-x86_64\" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified\nForced through blocking failures: Multiple precondition checks failed:\n* Precondition \"ClusterVersionUpgradeable\" failed because of \"ResourceDeletesInProgress\": Cluster minor level upgrades are not allowed while resource deletions are in progress; resources=clusterrolebinding \"default-account-openshift-machine-config-operator\"\n* Precondition \"EtcdRecentBackup\" failed because of \"ControllerStarted\": RecentBackup: The etcd backup controller is starting, and will decide if recent backups are available or if a backup is required\n* Precondition \"ClusterVersionRecommendedUpdate\" failed because of \"UnknownUpdate\": RetrievedUpdates=False (VersionNotFound), so the recommended status of updating from 4.14.18 to 4.15.5 is unknown.",
  "completionTime": "2024-03-25T03:31:22Z",
  "image": "quay.io/openshift-release-dev/ocp-release:4.15.5-x86_64",
  "startedTime": "2024-03-25T02:35:06Z",
  "state": "Completed",
  "verified": false,
  "version": "4.15.5"
}
{
  "acceptedRisks": "Target release version=\"\" image=\"quay.io/openshift-release-dev/ocp-release:4.14.18-x86_64\" cannot be verified, but continuing anyway because the update was forced: release images that are not accessed via digest cannot be verified\nForced through blocking failures: Multiple precondition checks failed:\n* Precondition \"ClusterVersionUpgradeable\" failed because of \"AdminAckRequired\": Kubernetes 1.27 and therefore OpenShift 4.14 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6958395 for details and instructions.\n* Precondition \"EtcdRecentBackup\" failed because of \"ControllerStarted\": RecentBackup: The etcd backup controller is starting, and will decide if recent backups are available or if a backup is required\n* Precondition \"ClusterVersionRecommendedUpdate\" failed because of \"UnknownUpdate\": RetrievedUpdates=False (VersionNotFound), so the recommended status of updating from 4.13.38 to 4.14.18 is unknown.",
  "completionTime": "2024-03-25T02:30:06Z",
  "image": "quay.io/openshift-release-dev/ocp-release:4.14.18-x86_64",
  "startedTime": "2024-03-25T01:29:39Z",
  "state": "Completed",
  "verified": false,
  "version": "4.14.18"
}
{
  "completionTime": "2024-03-25T01:06:12Z",
  "image": "quay.io/openshift-release-dev/ocp-release@sha256:db7a6fad1d56cf391159d252daabbb44178132081e077ae290317b91633fb1db",
  "startedTime": "2024-03-25T00:43:52Z",
  "state": "Completed",
  "verified": false,
  "version": "4.13.38"
}
```
check pending CSRs.
```console
$ oc get csr
NAME                                             AGE     SIGNERNAME                                    REQUESTOR                                                                         REQUESTEDDURATION   CONDITION
csr-5mgb5                                        4h11m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper         <none>              Approved,Issued
csr-6d8vj                                        4h15m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper         <none>              Approved,Issued
csr-8gczf                                        178m    kubernetes.io/kube-apiserver-client           system:node:ip-10-0-218-165.us-east-2.compute.internal                            24h                 Approved,Issued
csr-9vzb2                                        4h11m   kubernetes.io/kubelet-serving                 system:node:ip-10-0-144-7.us-east-2.compute.internal                              <none>              Approved,Issued
csr-bmhs7                                        4h17m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper         <none>              Approved,Issued
csr-d6sp6                                        178m    kubernetes.io/kube-apiserver-client           system:node:ip-10-0-150-183.us-east-2.compute.internal                            24h                 Approved,Issued
csr-d8dhx                                        4h15m   kubernetes.io/kubelet-serving                 system:node:ip-10-0-150-183.us-east-2.compute.internal                            <none>              Approved,Issued
csr-dwtqk                                        178m    kubernetes.io/kube-apiserver-client           system:node:ip-10-0-174-177.us-east-2.compute.internal                            24h                 Approved,Issued
csr-fgg77                                        4h5m    kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper         <none>              Approved,Issued
csr-g9dsm                                        178m    kubernetes.io/kube-apiserver-client           system:node:ip-10-0-174-23.us-east-2.compute.internal                             24h                 Approved,Issued
csr-jmlsm                                        4h15m   kubernetes.io/kubelet-serving                 system:node:ip-10-0-218-165.us-east-2.compute.internal                            <none>              Approved,Issued
csr-mkfd5                                        4h15m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper         <none>              Approved,Issued
csr-n74ww                                        4h3m    kubernetes.io/kubelet-serving                 system:node:ip-10-0-174-177.us-east-2.compute.internal                            <none>              Approved,Issued
csr-q8zlz                                        4h16m   kubernetes.io/kubelet-serving                 system:node:ip-10-0-174-23.us-east-2.compute.internal                             <none>              Approved,Issued
csr-x2nxr                                        178m    kubernetes.io/kube-apiserver-client           system:node:ip-10-0-144-7.us-east-2.compute.internal                              24h                 Approved,Issued
system:openshift:openshift-authenticator-7rp5m   4h15m   kubernetes.io/kube-apiserver-client           system:serviceaccount:openshift-authentication-operator:authentication-operator   <none>              Approved,Issued
system:openshift:openshift-monitoring-2hlg9      4h14m   kubernetes.io/kube-apiserver-client           system:serviceaccount:openshift-monitoring:cluster-monitoring-operator            <none>              Approved,Issued
system:openshift:openshift-monitoring-ldv6v      3h6m    kubernetes.io/kube-apiserver-client           system:serviceaccount:openshift-monitoring:cluster-monitoring-operator            <none>              Approved,Issued
```
scale up machineset to provision new node
```console
$ oc scale --replicas=2 machineset/rioliu-0325a-sngjv-worker-us-east-2a -n openshift-machine-api
machineset.machine.openshift.io/rioliu-0325a-sngjv-worker-us-east-2a scaled

$ machineset
NAME                                   DESIRED   CURRENT   READY   AVAILABLE   AGE
rioliu-0325a-sngjv-worker-us-east-2a   2         2         2       2           4h24m
rioliu-0325a-sngjv-worker-us-east-2b   1         1         1       1           4h24m
rioliu-0325a-sngjv-worker-us-east-2c   0         0                             4h24m
```
new machine can join the cluster successfully

**Comment by yuqi-zhang** [comment]
Thanks for the tests @rioliu-rh ! I think based on the testing results, it's at least no worse than what we had before. I think we can go ahead with the fix and see if this helps really old clusters (such as build02 which is actually having CSR issues)

File: pkg/daemon/certificate_writer.go [hunk 1]
```diff
@@ -340,41 +339,43 @@ func (dn *Daemon) syncControllerConfigHandler(key string) error {
...
+				return err
+			}
+
+			if err := runCmdSync("systemctl", "daemon-reload"); err != nil {
+				return err
 			}
-			if !bytes.Equal(currentKC.Clusters[0].Cluster.CertificateAuthorityData, data) {
-				return false, errors.New("cert data is not equal")
+
+			if err := runCmdSync("systemctl", "restart", "kubelet"); err != nil {
```
Comment (sinnykumari) [inline pkg/daemon/certificate_writer.go:376]: should this be start instead because we already stop kubelet service earlier?

**Review by sinnykumari — COMMENTED** [review]
One minor nit, other than that lgtm based on my limited context for this issue. I can add lgtm unless we want more set of eyes on the PR.

**Comment by sinnykumari** [comment]
/lgtm
putting hold for qe-approved label
/hold

**Comment by sinnykumari** [comment]
Thanks Rio for adding the stamp. Removing hold to get it merged.
/hold cancel

**Pull request merged**